File: Engine/src/main/java/org/etieskrill/engine/time/StepTimer.java
package org.etieskrill.engine.time;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class StepTimer {

    private long time = 0;
    private final Logger targetLogger;

    private static final Logger logger = LoggerFactory.getLogger(StepTimer.class);

    public StepTimer() {
        this(logger);
    }

    public StepTimer(Logger logger) {
        this.targetLogger = logger;
    }

    public void start() {
        time = System.nanoTime();
    }

    public void log(String message) {
        debug(message);
    }

    public void debug(String message) {
        targetLogger.debug("{} [{}s]", message, getTimeFormatted());
    }

    public void info(String subject) {
        targetLogger.info("{} [{}s]", subject, getTimeFormatted());
    }

    public void trace(String subject) {
        targetLogger.trace("{} [{}s]", subject, getTimeFormatted());
    }

    private double getTime() {
        long now = System.nanoTime();
        double delta = (now - time) / 1_000_000_000d;
        time = now;
        return delta;
    }

    private String getTimeFormatted() {
        return String.format("%.3f", getTime());
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/time/StepTimer.kt
package org.etieskrill.engine.time

import io.github.oshai.kotlinlogging.KLogger
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlin.time.DurationUnit
import kotlin.time.TimeSource

private val logger = KotlinLogging.logger {}

class StepTimer(
    val targetLogger: KLogger = logger,
) {

    private lateinit var time: kotlin.time.TimeMark

    fun start() {
        time = TimeSource.Monotonic.markNow()
    }

    fun log(message: () -> String) = debug(message)
    fun info(message: () -> String) = targetLogger.info { "${message()} [${getTimeFormatted()}s]" }
    fun debug(message: () -> String) = targetLogger.debug { "${message()} [${getTimeFormatted()}s]" }
    fun trace(message: () -> String) = targetLogger.trace { "${message()} [${getTimeFormatted()}s]" }

    fun getTimeFormatted(): String {
        return "%.3f".format(time.elapsedNow().toDouble(DurationUnit.SECONDS))
    }

}
